### PR TITLE
Integration of v8 segmentation

### PR DESCRIPTION
**Pull request by Laughing-q**
@AyushExel @glenn-jocher 
- add v8 segmentaion
    - training
    - evaluation
    - predict
- clean up a little bit
- simplify the outputs of Detect and Segment
- fix python interface issue by [https://github.com/ultralytics/ultralytics/issues/55#issuecomment-1366389721](https://github.com/ultralytics/ultralytics/issues/55#issuecomment-1366389721)

Verified training and predicting on balloon dataset, all works correctly to me.
results
3825919971_93fb1ec581_b
test command:
```python
# training
yolo task=segment mode="train" model=yolov8n-seg.yaml data=runs/data/balloon.yaml batch_size=16 epochs=50 device=\'0\' exist_ok=True plots=True v5loader=False
# predict
yolo task=segment mode=predict model=./runs/train/exp/weights/best.pt source=./dataset/balloon/images/val conf_thres=0.4
```

Also tested on COCO dataset to check training speed, it also looks fine to me.

Model info:
```vim
n vs n-seg: 8.9GFLOPs vs 11.9GFLOPs, 3.16M vs 3.39M
s vs s-seg: 28.8GFLOPs vs 39.5GFLOPs, 11.1M vs 11.7M
m vs m-seg: 79.3GFLOPs vs 103GFLOPs, 25.9M vs 27.1M
l vs l-seg: 165.7GFLOPs vs 207GFLOPs, 43.7M vs 45.7M
x vs x-seg: 258.5GFLOPs vs 324GFLOPs, 68.3M vs 71.4M
```
It's ready to merge after your review. :)
**BTW the segmentation training has to go with `v5loader=False`.**

## 🛠️ PR Summary

<sub>Made with ❤️ by [Ultralytics Actions](https://github.com/ultralytics/actions)<sub>

### 🌟 Summary
Enhancements to models, tests, and loss functions for segmentation in Ultralytics' AI framework.

### 📊 Key Changes
- Introduced Flops test to compare model complexities.
- Improved neural network weights handling in `.gitignore`.
- Made significant changes in the `Proto` and `Segment` modules, such as using `ConvTranspose2d` for upsampling in segmentation models.
- Integrated `Detect` and `Segment` into YOLO models with support for new architectures (e.g., `yolov8s-seg.yaml`).
- `check_model` assertion after ONNX export.
- Modified `neural architecture search` (NAS) in task-specific training scripts for segmentation.
- Added new segmentation loss functions and evaluation criteria.
- Updated configurations and utilities to support changes in the segmentation tasks and model definitions.

### 🎯 Purpose & Impact
- An increase in model architecture variety can potentially improve performance and accuracy for segmentation tasks in different use cases.
- Enhanced loss functions and evaluation metrics aim to improve model training feedback and accuracy.
- Updated configurations streamline the user experience when training or evaluating different YOLO models.
- Codebase maintainability is improved by ignoring model weights in Git and adding model checks after ONNX exports.

### CODE REVIEW

**Comment by Robotatron** [comment]
Is it intended for the segmentation integration to not support: 1) model ensembling 2) TTA (-- augment flag) 3) ClearML integration?

If so, is there any ETA to port those features from detection to segmentation?

**Comment by AyushExel** [comment]
Clearml integration it already supported for Segmentation

**Comment by Robotatron** [comment]
> Clearml integration it already supported for Segmentation

Strange. I followed the guide for using ClearML with YOLO5, did `clearml-init` and started training -> nothing was captured with ClearML. I then looked at the source code for segmentation loggers and only `comet` was supported, not ClearML.
Also the person who pushed the PR with ClearML integration acknowledges his PR was only for detection, not segmentation.

Did you try to use ClearML with segmentation yourself?

**Comment by AyushExel** [comment]
@Robotatron I think you're confused between 2 repos. The project that you linked is yolov5. this is Ulralytics. Here you can train with clearml enabled and all your metrics should be captured let me know if they aren't

**Comment by Robotatron** [comment]
> @Robotatron I think you're confused between 2 repos. The project that you linked is yolov5. this is Ulralytics. Here you can train with clearml enabled and all your metrics should be captured let me know if they aren't

I am confused indeed. Is Ultralytics [Ultralytics YOLOv8](https://github.com/ultralytics/ultralytics) a succesor / a better version of Ultralytics YOLOv5? I thought Ultralytics only had one product -> the YOLOv5

**Comment by AyushExel** [comment]
Yes its a successor and you can train even train yolov5 models with it
